TEIID-4594 Parquet Translator

aditya300899 opened the pull request:
This is the initial implementation of the execution logic for the parquet files. I think there should be a good amount of refinement required which can be done during testing as the java API docs for parquet don't explain anything, thus a little of the code is written with intuition. Kindly check!

rareddy commented on connectors/file/translator-parquet/src/main/java/org/teiid/translator/parquet/ParquetExecution.java:
```diff
@@ -0,0 +1,88 @@
+    }
+
+    List<Object> projectRow(Group row) {
+        ArrayList<Object> output = new ArrayList<Object>(this.visitor.getProjectedColumns().size());
```
I believe logic needs to match to what is being asked of projected columns, but below looks like as `select *`, i.e. projected columns can be just a single column or two or may be all, your output should match to that not all in parquet record

aditya300899 replied:
Yeah, it clicked me later, what is meant by the projected columns. I'll correct it.

aditya300899 replied:
Yeah, it clicked me later, what is meant by the projected columns. I'll correct it.

rareddy commented on connectors/file/translator-parquet/src/main/java/org/teiid/translator/parquet/ParquetExecutionFactory.java:
```diff
@@ -0,0 +1,91 @@
+    }
+
+    @Override
+    public boolean supportsCompareCriteriaEquals() {
```
I do think you supporting any of these below "supports" capabilities, these can be removed, you do not support `where` clause or `limit/offset`  or `in` which is what below are saying. Maybe there a way to extend current one to do that in that case keep it, next step could be that.

aditya300899 replied:
Yes, I just kept them as it is from the excel translator to figure out later if we support these.

aditya300899 replied:
Yes, I just put them as it is from excel to later figure out if we support these.

shawkins replied:
Go ahead and remove the supports methods for now.

rareddy commented on connectors/file/translator-parquet/src/main/java/org/teiid/translator/parquet/ParquetUpdateExecution.java:
```diff
@@ -0,0 +1,77 @@
+ * limitations under the License.
+ */
+
+package org.teiid.translator.parquet;
```
Are we supporting updates and deletes with Paraquet? typically with file-based we do not support these. In that case this class goes away.

aditya300899 replied:
There's an ExcelUpdateExecution in excel. Thus I just put a class in there. I don't know yet if these are to be supported in Parquet. You guys can tell better if it should be.

aditya300899 replied:
There's an ExcelUpdateExecution in excel. Thus I just put a class in there. I don't know yet if these are to be supported in Parquet. You guys can tell better if it should be.

aditya300899 commented on connectors/file/translator-parquet/src/main/java/org/teiid/translator/parquet/BaseParquetExecution.java:
```diff
@@ -0,0 +1,199 @@
+           File localFile = createTempFile(parquetFileStream);
+           Path path = new Path(localFile.toURI());
+           String extension = getExtension(localFile.getName());
+           if (extension != "parquet") {
```
Got it!

aditya300899 replied:
Got it!

aditya300899 commented on connectors/file/translator-parquet/src/main/java/org/teiid/translator/parquet/BaseParquetExecution.java:
```diff
@@ -0,0 +1,199 @@
+    }
+
+    public Group nextRow() throws TranslatorException, DataNotAvailableException {
+        while(true){
```
There's none. I took the inspiration from the excel one, but there the row was getting checked with the allow method and looped until a legit row was obtained. There's no need for it here. I got it wrong. Thanks for pointing out.

aditya300899 replied:
There's none. I took the inspiration from the excel one, but there the row was getting checked with the allow method and looped until a legit row was obtained. There's no need for it here. I got it wrong. Thanks for pointing out.

aditya300899 commented on connectors/file/translator-parquet/src/main/java/org/teiid/translator/parquet/ParquetExecution.java:
```diff
@@ -0,0 +1,88 @@
+                        output.add(row.getDouble(field, index));
+                        break;
+                    case "BYTE_ARRAY":
+                        output.add(row.getBinary(field,index));
```
done!

aditya300899 commented:
@shawkins you may have a look now!

shawkins commented:
@aditya300899  See https://github.com/shawkins/teiid/commits/aditya300899-Parquet-translator the main issue was needing to account for the page row count to prevent further reads.  It also seems like a good idea to filter to only parquet files - on my system the other spark files would cause exceptions.

aditya300899 commented:
> @aditya300899 See https://github.com/shawkins/teiid/commits/aditya300899-Parquet-translator the main issue was needing to account for the page row count to prevent further reads. 

Yeah, I see what I did there, I read the next row group and then put a check for null, but then I assigned the pages instance to again the next row group which was reading the next row group that was apparently null. Thanks a lot for the help. Now I see that it is returning the long and byte arrays, I'll see how to manage this tomorrow. 

> It also seems like a good idea to filter to only parquet files - on my system the other spark files would cause exceptions.
 
Yeah, I had it in my notes to be handled later but looks like you've done this except for the first file. Thanks!!

aditya300899 commented:
- We can manage different logical types in the way I've done here for managing a string by getting the logical type annotation. I'm skeptical if we have to manage all types of logical types mentioned in https://github.com/apache/parquet-format/blob/master/LogicalTypes.md in some way or not? 

- How are the expected column types to be dealt, in a similar way as done in the excel translator where they are converted from excel types? Also, I don't see such a check in the case of some translators like Cassandra. 

Also, an insight on how to wrap this part of the translator and how to move forward would be of great help!

shawkins commented:
> We can manage different logical types in the way I've done here for managing a string by getting the logical type annotation. I'm skeptical if we have to manage all types of logical types mentioned in https://github.com/apache/parquet-format/blob/master/LogicalTypes.md in some way or not?

You should add logical type mapping for everything that maps to a Teiid built-in type: http://teiid.github.io/teiid-documents/master/content/reference/r_supported-types.html

It also looks like the notion of repeated columns overlaps with using annotations to declare list/map values - you'll need to handle at list list types by mapping to array values.

> How are the expected column types to be dealt, in a similar way as done in the excel translator where they are converted from excel types? Also, I don't see such a check in the case of some translators like Cassandra.

The excel logic is a little general and does allow for the source ddl to have additional conversions over what is actually in the sheet.  Following the Cassandra logic is appropriate - the translator's primary job is to convert the source values into types known to Teiid - which are mostly the java sql types and some additional handling for things like arrays, java Date, byte[], and InputStreamFactory.  If the translator returns any other value object the engine will throw an exception that it can't be converted into a known runtime type.

aditya300899 commented:
@shawkins I've added support for a list and handled INT96 to return a BigInteger using a byte array. I have also included a test for the same. Can you check and let me know if we can move forward to the partitioning part now?

 However, like you said we would have to find SERDE lib or implement something of our own for a concrete solution.

shawkins commented on connectors/file/translator-parquet/src/main/java/org/teiid/translator/parquet/ParquetQueryVisitor.java:
```diff
@@ -0,0 +1,62 @@
+    public static int ROW_ID_INDEX = -1;
+
+    protected Stack<Object> onGoingExpression = new Stack<Object>();
+    private List<Integer> projectedColumns = new ArrayList<Integer>();
```
This is not yet wired in.  The projected columns should be determined from visiting the DerivedColumns.  You can either determine a projection index like Excel or just track the Columns that are being asked for and pass that to the execution to do further analysis.

aditya300899 replied:
I have removed the support methods for now. I would do the projectedColumns part at the earliest.

shawkins commented on connectors/file/translator-parquet/src/main/java/org/teiid/translator/parquet/BaseParquetExecution.java:
```diff
@@ -0,0 +1,189 @@
+           Configuration config = new Configuration();
+           reader = ParquetFileReader.open(HadoopInputFile.fromPath(path, config));
+           schema = reader.getFooter().getFileMetaData().getSchema();
+           columnIO = new ColumnIOFactory().getColumnIO(schema);
```
Once you have the expected columns, you can use that to read with a filtered schema.  That is you will construct a new MesssageType here that contains only fields that have been requested.  See the ColumnIOFactory.getColumnIO that take a requested schema argument.  I would guess that you would pass the filtered schema to the GroupRecordConverter as well.

aditya300899 replied:
I'll check this out! Thanks!

shawkins commented:
> Can you check and let me know if we can move forward to the partitioning part now?

Let's see if we can get the projection logic fully wired in and the extra logic pruned before moving on to partitioning.

aditya300899 commented on connectors/file/translator-parquet/src/main/java/org/teiid/translator/parquet/ParquetExecutionFactory.java:
```diff
@@ -0,0 +1,91 @@
+    }
+
+    @Override
+    public ParquetUpdateExecution createUpdateExecution(Command command,
```
Removed!

aditya300899 commented on connectors/file/translator-parquet/src/main/java/org/teiid/translator/parquet/ParquetExecution.java:
```diff
@@ -0,0 +1,128 @@
+                    continue;
+                }
+            } catch (NullPointerException e) {
+                //continue to check for primitive types
```
fieldType.getLogicalTypeAnnotation() gives null for non logical types like int64 and .toString throws the exception.

shawkins replied:
It would be best to have a null check rather than to catch the NPE as that may mask other logic errors.

aditya300899 commented on connectors/file/translator-parquet/src/main/java/org/teiid/translator/parquet/ParquetExecution.java:
```diff
@@ -0,0 +1,128 @@
+        for (int field = 0; field < fieldCount; field++) {
+            Type fieldType = row.getType().getType(field);
+            int valueCount = row.getFieldRepetitionCount(field);
+            if(valueCount > 1){
```
Yes,  I didn't think of it to be one, but it does need to be handled. I can't find any doc, so played a little. I observed that the valuecount comes out to be null for non-repeated field. In case of repeated fields, I think it is only possible with the non-annotated types.

aditya300899 commented on connectors/file/translator-parquet/src/main/java/org/teiid/translator/parquet/ParquetExecution.java:
```diff
@@ -0,0 +1,128 @@
+        return output;
+    }
+
+    private ArrayList getList(Group group) throws TranslatorException {
```
Do you mean array of objects or what? I don't quite follow here.

shawkins replied:
An Object array is fine.  The engine will convert to the internal representation from there.

aditya300899 commented:
>Let's see if we can get the projection logic fully wired in and the extra logic pruned before moving on to partitioning.

Yes, We should get done with this part first. I don't think this would take much time!

aditya300899 commented:
- I've added the filtered schema part on the basis of projected columns
- Also, I've added the implementation for non annotated repeated types, this needs to be tested as I did this on the basis of observation and analogy
-And the getList and getRepeatedList methods now return ArrayImpl instance which gives the desired outputs

@shawkins have a look!

shawkins commented:
> @shawkins have a look!

Looks like you are just about there.  Left a couple of comments about simplifications.  After that let's get you started on partitioning.

aditya300899 commented:
I don't see any comments @shawkins

aditya300899 commented on connectors/file/translator-parquet/src/main/java/org/teiid/translator/parquet/ParquetExecution.java:
```diff
@@ -0,0 +1,150 @@
+                output.add(null);
+            }
+            PrimitiveType.PrimitiveTypeName type = fieldType.asPrimitiveType().getPrimitiveTypeName();
+            for (int index = 0; index < valueCount; index++) {
```
I was thinking to remove this as well.

aditya300899 commented on connectors/file/translator-parquet/src/main/java/org/teiid/translator/parquet/ParquetExecution.java:
```diff
@@ -0,0 +1,150 @@
+    }
+
+    Object getPrimitiveTypeObject(PrimitiveType.PrimitiveTypeName primitiveTypeName, int field, int index, Group row, Type fieldType) {
+        switch (primitiveTypeName) {
```
Yes, it needs to be. Can we also return this as binary? The only difference between binary and FIXED_LEN_BYTE_ARRAY is that the latter is of fixed length and the former of variable.

shawkins replied:
You can return both as byte arrays.  We only support a varbinary type and not a fixed binary type.

aditya300899 commented on connectors/file/translator-parquet/src/main/java/org/teiid/translator/parquet/ParquetQueryVisitor.java:
```diff
@@ -0,0 +1,88 @@
+
+    private void createProjectedColumn() {
+        Column column = (Column) this.onGoingExpression.pop();
+        String str = column.getProperty(ParquetMetadataProcessor.COLUMN_NUMBER, false);
```
Yes, this looks better. I've implemented this. Thanks!

aditya300899 commented:
@shawkins I think I've implemented the asked changes. Please have a look!

shawkins commented:
> I think I've implemented the asked changes. Please have a look!

Excellent.  This constitutes a milestone - full pushdown support of projection.  Once we understand more about the requirements of partitioning - especially in terms of the expectation of the file / location option - we'll be ready to merge this and keep refining from there.

Did you find a partitioning scenario to start with?

aditya300899 commented:
> Did you find a partitioning scenario to start with?

I'm looking into this.

aditya300899 commented:
@shawkins I've refined the initial directory-based partitioning! Please have a look! And leave a comment on the next step. Thanks

shawkins commented on connectors/file/translator-parquet/src/main/java/org/teiid/translator/parquet/ParquetQueryVisitor.java:
```diff
@@ -0,0 +1,112 @@
+
+    @Override
+    public void visit(Comparison obj) {
+        if(this.table.getProperty(ParquetMetadataProcessor.PARTITIONING_SCHEME) == null){
```
you can remove this consideration from here and just inline the directoryBasedPartitioning method - you are simply trying to build the column / value map, then handle that after the visitor run.

aditya300899 replied:
What I understand from this is that we might use the map even for non-partitioned column comparisons and thus there is no need for that, right?

shawkins replied:
Yes that is correct.  You can handle how the predicates are supposed to be interpreted after visitation and that will keep the visitor logic much simpler.

shawkins commented on connectors/file/translator-parquet/src/main/java/org/teiid/translator/parquet/ParquetQueryVisitor.java:
```diff
@@ -0,0 +1,112 @@
+        Column column = (Column)this.onGoingExpression.pop();
+        String columnName = column.getSourceName();
+        visitNode(obj.getRightExpression());
+        String columnValue = (String) this.onGoingExpression.pop();
```
Thinking beyond the partitioning columns it is not required that the literal value will be a String.  It can be any of the comparable value types that are supported by the translator.  For the partitioning columns we would require that the type value is easily convertable to string - for example exact numeric would be allowable as well.  It's important to have the types represented correctly as there is a semantic difference between string an numeric types that comes into play once you support other comparisons.  For example with 10 > 2 as an integer comparison, but '10' < '2' as a string comparison.

Also, since the visitor is performing casts in any case, you may find it more straight-forward to do something like:

public void visit(Comparison obj) {
  ColumnReference cr = (ColumnReference)obj.getLeftExpression();
  Column column = cr.getMetadataObject();
  Literal l = (Literal)obj.getRightExpression();
  ...
}

The ongoing expression logic will only be helpful if there ends up being handling that is common for very large expressions / subtrees.

If you want to consider the other comparisons and IN predicates at this point you also want to make what you are capturing pretty general - it will need more than just the column and the value to also convey the type of operation (for example less than vs. equality) and a set of values (IN ...)

aditya300899 replied:
> Thinking beyond the partitioning columns it is not required that the literal value will be a String.

Yes, I initially took it to be just the directory partitioning where the user can be asked to name the column as strings. But since we would also need to handle other aspects and comparison types other than equality, this would fail, we need to know the types of values too. What all types are we going to handle? Also, what datatypes would allow what kind of comparisons is also a question. For eg, I don't think a complex data type like an array would be supported. Isn't it?

> The ongoing expression logic will only be helpful if there ends up being handling that is common for very large expressions / subtrees.

Right! The visiting of them is not needed, I was also wondering if there's a need for this visiting even in the excel query visitor. That also seems similar and could be simplified like you mentioned. 

>If you want to consider the other comparisons and IN predicates at this point you also want to make what you are capturing pretty general - it will need more than just the column and the value to also convey the type of operation (for example less than vs. equality) and a set of values (IN ...)

I think I should rather implement all of it together which would make the capability wholesome and things can be implemented accordingly. Otherwise, like you said they would have to be generalized or changed anyway in the future when the support is extended. What do you say?

shawkins replied:
>  What all types are we going to handle? Also, what datatypes would allow what kind of comparisons is also a question. For eg, I don't think a complex data type like an array would be supported. Isn't it?

Generally the translator will need to support literal values that match the return values to the engine.  In this case you have string, varbinary, biginteger, long, integer, float, double, boolean - all of those are comparable types that can end up in a predicate of the form "col = value literal".  However partitioning columns may need additional restrictions, such as not using varbinary, float, double values - as those could have ambiguous string values.  

> Otherwise, like you said they would have to be generalized or changed anyway in the future when the support is extended. What do you say?

It will be more difficult to fully implement everything at once, but you can make additional logic easier just by capturing more information, rather than less, in your visitation / query analysis.  Some things will be pretty easier to layer in, such as once you support predicates on non-partitioning columns you'll need to add the logic to convert those predicates into a filter.  Other things will be more difficult, such as once you support comparisons other than equality, then you can't simply build up a file pattern to honor the predicate - you'd have to implement a check against the file path - which will also require updating the VirtualFile api to expose both a file name and a file path.

aditya300899 replied:
> However partitioning columns may need additional restrictions, such as not using varbinary, float, double values - as those could have ambiguous string values.

So, I'm implementing the visitor as something that builds a map on the HashMap<String(ColumnName), Literal>. So when I get the literal corresponding to a column, I would implement it using a switch case using the class name(Is there an enum for the same, I mean an API that would give us the teiid types for a corresponding literal) that would convert the respective string,  biginteger, long, integer,  boolean to strings using the toString methods and default the switch case to throw an exception for incompatible types for a directory path like the varbinary, float, and double. Thoughts?

>  It will be more difficult to fully implement everything at once

Alright, we'll do things one step at a time and make progress that way. 

> you'd have to implement a check against the file path - which will also require updating the VirtualFile api to expose both a file name and a file path.

Right!!

shawkins replied:
> Thoughts?

Yes that is the right track.  Capture the full literal, or even the full predicate, for further work.

Also note that opening up to additional comparisons allows for thing like "between" scenarios - where col > 'value1' and col < 'value2' - which will further complicate handling.

aditya300899 replied:
> or even the full predicate

Like just store the complete comparison object, and later use it for different purposes, like here we can probably utilize it directly to build the directory path, right?

> Also note that opening up to additional comparisons allows for thing like "between" scenarios - where col > 'value1' and col < 'value2' - which will further complicate handling.

This would be after the virtualfile API is capable to give the file path, right?

shawkins replied:
> Like just store the complete comparison object, and later use it for different purposes, like here we can probably utilize it directly to build the directory path, right?

Yes a map of columns to predicates would be the most general.

> This would be after the virtualfile API is capable to give the file path, right?

Yes, and I'm trying to highlight that there could be more than one predicate for a given column.  That can be handled by a multi-map or similar.

shawkins commented on connectors/file/translator-parquet/src/main/java/org/teiid/translator/parquet/BaseParquetExecution.java:
```diff
@@ -0,0 +1,226 @@
+    }
+
+    private String getDirectoryPath(HashMap<String, String> columnValues, String partitionedColumns) {
+      String[] partitionedColumnsArray = getPartitionedColumns(partitionedColumns);
```
Are you basing your example off of default spark partitioning?  We've also seen the convention where the directory structure contains the name and value:

dir/dim1=value1/dim2=value2/*

vs here we have

dir/value1/value2/*

 Do you seen any taxonomy or template for how the partitioning scheme is capture in the hive metastore?

aditya300899 replied:
The spark used the "=" i.e of the form dir/dim1=value1/dim2=value2/*,  according to https://spark.apache.org/docs/2.4.0/sql-data-sources-parquet.html.

What I saw from a link, in hive partitioning looks like dir/value1/value2/*, see https://data-flair.training/blogs/apache-hive-partitions/ 

Then I checked a few pages today, one of then shows that it uses the same pattern dir/dim1=value1/dim2=value2/*, see https://www.guru99.com/hive-partitions-buckets-example.html

I think this dir/dim1=value1/dim2=value2/* should be considered, what do you say?

shawkins replied:
> I think this dir/dim1=value1/dim2=value2/* should be considered, what do you say?

Yes it should be considered.  From what I can gather that is the dominant paradigm.

aditya300899 commented:
@shawkins this is what I've come up with, this might need some more refinements which can be handled as we progress. Let me know if this looks okay to you or if there are any changes! Thanks!

shawkins commented on connectors/file/translator-parquet/src/main/java/org/teiid/translator/parquet/BaseParquetExecution.java:
```diff
@@ -0,0 +1,258 @@
+            if (pageRowCount-- <= 0) {
+                this.rowIterator = null;
+                PageReadStore nextRowGroup = this.reader.readNextRowGroup();
+                if(nextRowGroup == null){
```
I didn't handle the case of nextRowGroup != null here.  You'll need to account for that by resetting the pageRowCount and the rowIterator.

shawkins commented on connectors/file/translator-parquet/src/test/java/org/teiid/translator/parquet/TestParquetExecution.java:
```diff
@@ -0,0 +1,100 @@
+        Mockito.stub(connection.getFiles("people1.parquet")).toReturn(JavaVirtualFile.getFiles("people1.parquet", new File(UnitTestUtil.getTestDataPath(), "people1.parquet")));
+
+        ArrayList results = helpExecute(ddl, connection, "select * from Table1");
+        Assert.assertEquals("[[Aditya, 1, Sharma], [Animesh, 2, Sharma], [Shradha, 3, Khapra]]", results.toString());
```
Be sure to add a test here in the non-partitioned case covering projecting a subset of columns.

aditya300899 commented:
@shawkins , Please review! Also, the direct method call doesn't work for me as well!

shawkins commented on connectors/file/translator-parquet/src/main/java/org/teiid/translator/parquet/ParquetPlugin.java:
```diff
@@ -0,0 +1,41 @@
+    public static final BundleUtil Util = new BundleUtil(PLUGIN_ID,PLUGIN_ID + ".i18n", ResourceBundle.getBundle(PLUGIN_ID + ".i18n")); //$NON-NLS-1$ //$NON-NLS-2$
+
+    public static enum Event implements BundleUtil.Event{
+        TEIID23000,
```
Are any of these in use?

aditya300899 replied:
Only TEIID23000. I haven't got the complete usage of this yet. It'd be great to know

aditya300899 commented on connectors/file/translator-parquet/src/main/java/org/teiid/translator/parquet/BaseParquetExecution.java:
```diff
@@ -0,0 +1,358 @@
+
+    private String[] getPartitionedColumns(String partitionedColumns) {
+        String[] partitionedColumnList;
+        partitionedColumnList = partitionedColumns.split(",");
```
How do we do this?

shawkins replied:
Capture initial limitations / exceptions on the jira so that it's easier to create the documentation.  Especially without a metadata importer we should capture:

- expectations for the partitioning property format
- type expectations for partitioning columns
- supported sources - it looks like path based partitioning will not yet work against ftp
- supported types / type mapping

aditya300899 commented on connectors/file/translator-parquet/src/main/java/org/teiid/translator/parquet/BaseParquetExecution.java:
```diff
@@ -0,0 +1,358 @@
+                              throw new TranslatorException("Only string, biginteger, long, integer, boolean are supported.");
+                      }
+                      break;
+                  default: throw new TranslatorException("Only equal comparison is allowed");
```
How about deleting only those elements in the map which have equality comparison and leave others for the filtering? This shall work if the partitioned columns are also in the file.

shawkins replied:
> This shall work if the partitioned columns are also in the file.

Can we confirm if this is generally the case?

If so, then yes you can absolutely just let them be filters instead of part of the path.

shawkins commented:
@aditya300899 can you also add a few more tests focusing on failing scenarios - cases where you throw an exception.  Either the enforce a hard expectation of an exception or they are great as a starting point later once someone wants to do further work from a "test first" perspective.

aditya300899 commented on connectors/file/translator-parquet/src/test/java/org/teiid/translator/parquet/TestParquetExecution.java:
```diff
@@ -0,0 +1,207 @@
+    }
+
+    @Test
+    public void testParquetExecutionWithRowFilterEq() throws Exception {
```
@shawkins this test is failing despite the correct rowGroupFilter, please have a look! It is not returning any row.

shawkins commented on connectors/file/translator-parquet/src/main/java/org/teiid/translator/parquet/BaseParquetExecution.java:
```diff
@@ -0,0 +1,338 @@
+        String[] partitionedColumnsArray = getPartitionedColumns(partitionedColumns);
+        for (String s : partitionedColumnsArray) {
+            partitionedColumnsHm.put(s, 1);
+            if (predicates.containsKey(s)) {
```
This needs refined a bit more:

path.append("/").append(s).append("="); 
String value = "*";
Comparison predicate = predicates.remove(s);
if (predicate != null) {
   ...
} 
path.append(value);

Also if partial filters is ever turned off, we'll need to eventually add escape handling to the path syntax just in case a string value contains *.

shawkins commented on connectors/file/translator-parquet/src/main/java/org/teiid/translator/parquet/BaseParquetExecution.java:
```diff
@@ -0,0 +1,338 @@
+        if(columnPredicates.size() == 0){
+            return FilterCompat.NOOP;
+        }
+        Iterator hashMapIterator = columnPredicates.entrySet().iterator();
```
Please use generic types when possible.  For example instead of just Iterator, use Iterator<Entry<String, Comparison>> - the ide should have an option to infer generic types as a formatting/cleanup task.

shawkins commented on connectors/file/translator-parquet/src/main/java/org/teiid/translator/parquet/BaseParquetExecution.java:
```diff
@@ -0,0 +1,338 @@
+        StringBuilder path = new StringBuilder(); // because we are only supporting the equality comparison as of now, otherwise we would return a list of paths to iterate over
+        String[] partitionedColumnsArray = getPartitionedColumns(partitionedColumns);
+        for (String s : partitionedColumnsArray) {
+            partitionedColumnsHm.put(s, 1);
```
Do you mean for this to always be a constant of 1?

aditya300899 replied:
Yes, this is only to check if the column is a partitioned one or not! I could have used other DS, but this seemed best optimal to me!

aditya300899 commented:
@shawkins I think it should be fine now. I'll add the limitations in the JIRA comment. 
Also, the build in the travis is failing at some other module, kindly check that.

shawkins commented:
> Also, the build in the travis is failing at some other module, kindly check that.

More than likely you have incompletely applied the changes from master - see that the master build has no issues.

shawkins commented:
> More than likely you have incompletely applied the changes from master - see that the master build has no issues.

Actually the changes I gave you weren't applied to master and thus were not yet complete - there are some minor adjustments for that test class still needed.

shawkins commented:
I'll commit this to master once the test run completes: https://github.com/shawkins/teiid/commit/ff07933de62b4d376cd223780d5ea28f2f69518c

Basically it just needs to have the getFile method changed to public.  You can make the same change locally or rebase again to master.

aditya300899 commented:
Thanks @shawkins. It's fine now!

shawkins commented on connectors/file/translator-parquet/src/main/java/org/teiid/translator/parquet/BaseParquetExecution.java:
```diff
@@ -0,0 +1,341 @@
+            partitionedColumnsHm.put(s, 1);
+            path.append("/").append(s).append("=");
+            String value = "*";
+            if (predicates.containsKey(s)) {
```
It's usually best to just combine contains and remove since remove returns the value.

shawkins commented on connectors/file/translator-parquet/src/main/java/org/teiid/translator/parquet/BaseParquetExecution.java:
```diff
@@ -0,0 +1,341 @@
+
+    private String getDirectoryPath(HashMap<String, Comparison> predicates, String partitionedColumns) throws TranslatorException {
+        StringBuilder path = new StringBuilder(); // because we are only supporting the equality comparison as of now, otherwise we would return a list of paths to iterate over
+        String[] partitionedColumnsArray = getPartitionedColumns(partitionedColumns);
```
A map with a constant value is effectively the same as a set.  You can change the type of partitionedColumnsHm to a HashSet or a LinkedHashSet if you want to preserve the order.

Also since it's not possible for partitionedColumnsHm to be null you can avoid null checks against it.

aditya300899 replied:
> Also since it's not possible for partitionedColumnsHm to be null you can avoid null checks against it.

I think it's possible when there's no partitioning.

shawkins replied:
It's currently initialized as an empty map and never set to null.

shawkins commented on connectors/file/translator-parquet/src/main/java/org/teiid/translator/parquet/BaseParquetExecution.java:
```diff
@@ -0,0 +1,341 @@
+            path.append(value);
+        }
+        path.append("/*");
+    return path.toString();
```
This needs indented a little more.

shawkins commented on connectors/file/translator-parquet/src/main/java/org/teiid/translator/parquet/BaseParquetExecution.java:
```diff
@@ -0,0 +1,341 @@
+            case LT: return "lt";
+            case GE: return "gtEq";
+            case LE: return "ltEq";
+            case NE: return "NotEq";
```
Capitalized NotEq does not appear to be correct - can you add a not equals test?

shawkins commented on connectors/file/translator-parquet/src/main/java/org/teiid/translator/parquet/ParquetQueryVisitor.java:
```diff
@@ -0,0 +1,98 @@
+    private List<String> projectedColumnNames = new ArrayList<String>();
+    protected ArrayList<TranslatorException> exceptions = new ArrayList<TranslatorException>();
+
+    public HashMap<String, Comparison> getColumnPredicates() {
```
Now that ordered predicates are supported the strategy of mapping to a single Comparison won't work.  For example add a test that has a predicate like WHERE id > 2 and id < 4.  You don't need to check for the case of an equality predicate and something else - the engine will always reduce down to a single equality predicate when possible.

aditya300899 replied:
> Now that ordered predicates are supported the strategy of mapping to a single Comparison won't work

Yeah, what do we need to do then, a multimap?

>You don't need to check for the case of an equality predicate and something else - the engine will always reduce down to a single equality predicate when possible.

Sorry. but I didn't get this. where am I checking for an equality predicate?

shawkins replied:
> Yeah, what do we need to do then, a multimap?

Yes a multi-map will work.

> Sorry. but I didn't get this. where am I checking for an equality predicate?

I just mean you don't have to do any additional special checks at the translator level or try to logically combine predicates.

aditya300899 commented on connectors/file/translator-parquet/src/main/java/org/teiid/translator/parquet/BaseParquetExecution.java:
```diff
@@ -0,0 +1,345 @@
+            path.append("/").append(s).append("=");
+            String value = "*";
+            Collection<Comparison> columnPredicates = predicates.get(s);
+            for(Comparison predicate: columnPredicates) {
```
we can just check for a single predicate here instead of a for loop, just did that to make it look cleaner.

shawkins commented:
I think that this can considered good for the initial merge.  I'll make a couple of refinements based upon TEIID-6025
